sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -533,6 +538,56 @@ public static Unbound withSideInputs(
     return DisplayData.item("fn", fn.getClass()).withLabel("Transform Function");
   }
 
+  private static void finishSpecifyingStateSpecs(
+      DoFn<?, ?> fn,
+      CoderRegistry coderRegistry,
+      Coder<?> inputCoder) {
+    DoFnSignature signature = DoFnSignatures.getSignature(fn.getClass());
+    Map<String, DoFnSignature.StateDeclaration> stateDeclarations = signature.stateDeclarations();
+    for (DoFnSignature.StateDeclaration stateDeclaration : stateDeclarations.values()) {
+      try {
+        StateSpec<?, ?> stateSpec = (StateSpec<?, ?>) stateDeclaration.field().get(fn);
+        stateSpec.offerCoders(codersForStateSpecTypes(stateDeclaration, coderRegistry, inputCoder));
+        stateSpec.finishSpecifying();
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      } catch (Exception e) {
+        throw new Pipeline.PipelineExecutionException(e);
```

Comment:
You're right, there are no checked exceptions. Implementations of `StateSpec.finishSpecifying` may throw `IllegalStateException`, but since we don't need to wrap them in `Pipeline.PipelineExecutionException` I can remove the second `catch` clause.